File: ros2/workspace/src/simple_ros_cpp/src/service_server.cpp
#include <iostream>
#include <chrono> 
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/set_bool.hpp"

using namespace std; 
void printHello(); 

typedef shared_ptr<std_srvs::srv::SetBool::Request> Req; 
typedef shared_ptr<std_srvs::srv::SetBool::Response> Res; 

class ServiceServer : public rclcpp::Node {
    public:
        ServiceServer() : Node("ServiceServer") {
            _server = create_service<std_srvs::srv::SetBool>(
                "setBoolService", 
                std::bind(&ServiceServer::service_callback, this, std::placeholders::_1, std::placeholders::_2)); 
                
            
            // _twist_pub_timer = create_wall_timer(1s, std::bind(&ServiceServer::twist_pub, this)); 
            // _update_timer = create_wall_timer(17ms, std::bind(&ServiceServer::update, this)); 
        }
    private:
        int _count; 
        bool isTrue = false; 
        rclcpp::Service<std_srvs::srv::SetBool>::SharedPtr _server; 
        rclcpp::TimerBase::SharedPtr _twist_pub_timer; 
        rclcpp::TimerBase::SharedPtr _update_timer; 
        
        void service_callback(const Req request, Res response) {
            RCLCPP_INFO(get_logger(), "setBoolService received " + request->data); 
            
            isTrue = request->data; 

            if (request->data ) {
                response->success = true; 
                response->message = "True response"; 
            } else {
                response->success = false; 
                response->message = "False response"; 
            }
            
        }
}; 

int main()
{
    rclcpp::init(0, nullptr);
    auto node = std::make_shared<ServiceServer>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2/workspace/src/simple_ros/simple_ros/service_server.py
import rclpy
from rclpy.node import Node
from rclpy.qos import (
    QoSDurabilityPolicy,
    QoSHistoryPolicy,
    QoSProfile,
    QoSReliabilityPolicy,
)
from std_msgs.msg import String
from std_srvs.srv import SetBool 
import time

class Service_server(Node):
    def __init__(self):
        super().__init__("Service_server")
        self.create_timer(1, self.print_hello)
        self.create_service(SetBool, "setBoolService", self.setBool_callback) 
        self.bool = bool()
        self.count = 0

    def setBool_callback(self, request:SetBool.Request, response:SetBool.Response):
        self.get_logger().info(f"request data : {request.data}")
        self.bool = request.data
        self.count += 1

        if request.data:
            response.success = True
            response.message = "True response"
        else :
            response.success = False
            response.message = "False response"

        self.get_logger().info(f"response {self.count} => {response.message}")

        return response

    def print_hello(self):
        #self.get_logger().info(f"current bool : {self.bool}")
        pass

def main():
    rclpy.init()
    node = Service_server()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info(f"KeyboardInterrupted")
    finally:
        node.destroy_node()

if __name__ == "__main__":
    main()
